packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -1868,6 +1871,212 @@ function commitPassiveLifeCycles(finishedWork: Fiber): void {
   }
 }
 
+function commitDoubleInvokeEffectsInDEV(
+  fiber: Fiber,
+  hasPassiveEffects: boolean,
+) {
+  if (enableDoubleInvokingEffects) {
+    if (__DEV__) {
+      invokeLayoutEffectsUnmountInDEV(fiber);
+      if (hasPassiveEffects) {
+        invokePassiveEffectsUnmountInDEV(fiber);
+      }
+
+      invokeLayoutEffectsMountInDEV(fiber);
+      if (hasPassiveEffects) {
+        invokePassiveEffectsMountInDEV(fiber);
+      }
+    }
+  }
+}
+
+function invokeLayoutEffectsUnmountInDEV(firstChild) {
+  if (enableDoubleInvokingEffects) {
+    if (__DEV__) {
+      // unmount layout effects
+      let fiber = firstChild;
+      while (fiber !== null) {
+        if (fiber.child !== null) {
+          // Should we add a separate subtree tag for this?
+          const primarySubtreeTag = fiber.subtreeTag & LayoutSubtreeTag;
+          if (primarySubtreeTag !== NoSubtreeTag) {
+            invokeLayoutEffectsUnmountInDEV(fiber.child);
+          }
+        }
+
+        const effectTag = fiber.effectTag;
+        const current = fiber.alternate;
+        // This is a mount
+        if (current === null) {
+          if (effectTag & Update) {
+            switch (fiber.tag) {
```

Comment:
I tried to clarify the split between CommitWork and WorkLoop a bit more in https://github.com/facebook/react/pull/19599

I think it's best summed up as the idea is that CommitWork should be kind of the work that each fiber does in the commit. Where as WorkLoop is about coordinating across the Fibers. (Possibly WorkLoop could be split into more files but CommitWork still be just the work that individual Fiber does.) Think of it as `fiber.commitWork()`.

Practically that means that these switch statements based on tag should be in functions within `commitWork` but the traversal should be in WorkLoop.

So I'd recommending splitting out just each of these switch statements into its own function. Let that function remain in this file. Then move the outer one back to WorkLoop.

Reply:
> I think it's best summed up as the idea is that CommitWork should be kind of the work that each fiber does in the commit. Where as WorkLoop is about coordinating across the Fibers.

Except for things like `commitNestedUnmounts` 😆